src/EditorFeatures/Core/Organizing/OrganizeDocumentCommandHandler.cs

```diff
@@ -115,61 +93,110 @@ private static bool IsCommandSupported(EditorCommandArgs args, bool needsSemanti
             return false;
         }
 
-        public bool ExecuteCommand(SortImportsCommandArgs args, CommandExecutionContext context)
+        private bool ExecuteCommand(
+            EditorCommandArgs commandArgs,
+            CommandExecutionContext context,
+            Func<ITextSnapshot, IUIThreadOperationContext, Task<Document?>> getCurrentDocumentAsync,
+            Func<Document, CancellationToken, Task<Document>> getChangedDocumentAsync)
         {
-            using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Organizing_document))
-            {
-                SortImports(args.SubjectBuffer, context.OperationContext);
-            }
+            // We're showing our own UI, ensure the editor doesn't show anything itself.
+            context.OperationContext.TakeOwnership();
+
+            var subjectBuffer = commandArgs.SubjectBuffer;
+            var textView = commandArgs.TextView;
+
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            if (caretPoint is null)
+                return false;
 
+            if (!subjectBuffer.TryGetWorkspace(out var workspace))
+                return false;
+
+            var token = _listener.BeginAsyncOperation(nameof(ExecuteCommand));
+
+            var snapshotSpan = textView.GetTextElementSpan(caretPoint.Value);
+            ExecuteAsync(workspace, commandArgs, snapshotSpan, getCurrentDocumentAsync, getChangedDocumentAsync)
+                .ReportNonFatalErrorAsync()
+                .CompletesAsyncOperation(token);
             return true;
         }
 
-        public bool ExecuteCommand(SortAndRemoveUnnecessaryImportsCommandArgs args, CommandExecutionContext context)
+        private async Task ExecuteAsync(
+            Workspace workspace,
+            EditorCommandArgs commandArgs,
+            SnapshotSpan snapshotSpan,
+            Func<ITextSnapshot, IUIThreadOperationContext, Task<Document?>> getCurrentDocumentAsync,
+            Func<Document, CancellationToken, Task<Document>> getChangedDocumentAsync)
         {
-            using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Organizing_document))
-            {
-                this.SortAndRemoveUnusedImports(args.SubjectBuffer, context.OperationContext);
-            }
+            _threadingContext.ThrowIfNotOnUIThread();
 
-            return true;
+            var indicatorFactory = workspace.Services.GetRequiredService<IBackgroundWorkIndicatorFactory>();
+            using var backgroundWorkContext = indicatorFactory.Create(
+                commandArgs.TextView,
+                snapshotSpan,
+                EditorFeaturesResources.Organizing_document,
+                cancelOnEdit: true,
+                cancelOnFocusLost: true);
```

Comment:
💭 Seems like this would be unnecessary

```suggestion
                cancelOnFocusLost: false);
```

Reply:
will do.